File: pat1020.py
"""
    思路：1.根据中序和后序找到根节点
          2.根据postion的值决定根节点在列表的位置，左子树的根节点为2*i+1，
            右子树的根节点为2*i+2
          3.将列表按序输出

"""


postorder = []
inorder = []
result = 100000 * ['0']   #   初始化列表，列表要足够大

def level_order(pos, postl, postr, inl, inr):
    if postl > postr:
        return 
    root = postorder[postr]
    r_index = inorder.index(root)     #   在中序序列找到root的索引
    result[pos] = root

    numleft = r_index - inl      #   左子树节点个数
    level_order(2 * pos + 1, postl, postl + numleft - 1, inl, r_index - 1)     #   对左子树层序遍历
    level_order(2 * pos + 2, postl + numleft, postr - 1, r_index + 1, inr)     #   对右子树层序遍历

N = int(input()) - 1
postorder = input().split()
inorder = input().split()

level_order(0, 0, N, 0, N)

print(result[0],end="")
cnt = 0
for x in result[1:]:
    if x != "0":
        cnt += 1
        print(" " + x,end="")
    if cnt == N:
        break

File: pat1020.cpp
/*
	思路：1.根据后序和中序确定根节点，递归解决
		  2.在递归的时候，顺便更新位置参数，根据2*i+1，和2*i+2决定层序遍历顺序
*/

#include<iostream>
using namespace std;

int result[1000000] = { 0 };		//	存放result的数组要足够大，否则第五点会出现段错误
int inorder[30];
int postorder[30];

void levelorder(int postion, int postl, int postr, int inl, int inr) {
	if (postl > postr) {
		return;
	}

	int root = postorder[postr];
	int i = inl;
	while (root != inorder[i]) {
		++i;
	}
	int numleft = i - inl;		//	左子树节点个数
	result[postion] = root;		//	将根节点按层序遍历的顺序放在result数组中

	levelorder(2 * postion + 1, postl, postl + numleft - 1, inl, i - 1);
	levelorder(2 * postion + 2, postl + numleft, postr - 1, i + 1, inr);
}

int main() {

	int N;
	cin >> N;
	for (int i = 0; i < N; i++) {
		scanf("%d", &postorder[i]);
	}
	for (int i = 0; i < N; i++) {
		scanf("%d", &inorder[i]);
	}
	levelorder(0, 0, N - 1, 0, N - 1);
	cout << result[0];
	for (int i = 1, cnt = 1; cnt < N; i++) {
		if (result[i] != 0) {
			cout << " " << result[i];
			++cnt;
		}
	}
	return 0;

}